### PR TITLE
[FEATURE] Integrate `DataContextVariables` with `DataContext`

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Enable `context.variables` syntax
- Replace all internal usage of `self.config` with `self.variables` to delegate responsibility over config management to the new object.
- Update existing tests to work with changes.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Comment by ghost** [comment]
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=5466&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=5466&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[Review these changes using an interactive CodeSee Map](https://app.codesee.io/r/reviews?pr=5466&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
CodeSee Map Legend

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -157,7 +158,9 @@ def __init__(self, runtime_environment: Optional[dict] = None) -> None:
         self._data_context_id = self._construct_data_context_id()
 
         # Override the project_config data_context_id if an expectations_store was already set up
-        self.config.anonymous_usage_statistics.data_context_id = self._data_context_id
+        self.variables.config.anonymous_usage_statistics.data_context_id = (
+            self._data_context_id
+        )
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: The config is managed by `self.variables`. As such, we want to make sure that any updates made to the config are reflected in the variable's attribute.

This PR refactors all usages of `self.config` to ensure there's no mismatch of state.

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 2]
```diff
@@ -250,7 +253,7 @@ def plugins_directory(self) -> Optional[str]:
         # NOTE: <DataContextRefactor>  Why does this exist in AbstractDataContext? CloudDataContext and
         # FileDataContext both use it. Determine whether this should stay here or in child classes
         return self._normalize_absolute_or_relative_path(
-            self.project_config_with_variables_substituted.plugins_directory
+            self.variables.plugins_directory
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py:254]: :bow: Cleaner now that the substitutions are in `DataContextVariables`

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 3]
```diff
@@ -357,7 +358,7 @@ def checkpoint_store(self) -> "CheckpointStore":  # noqa: F821
     @property
     def profiler_store_name(self) -> Optional[str]:
         try:
-            return self.project_config_with_variables_substituted.profiler_store_name
+            return self.variables.profiler_store_name
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py:359]: Should this also create a profiler if `profiler_store_name` is None?

File: great_expectations/data_context/data_context/base_data_context.py [hunk 4]
```diff
@@ -308,6 +308,9 @@ def __init__(
         # It is rather clunkly and we should explore other ways of ensuring that BaseDataContext has all of the
         # necessary properties / overrides
         self._synchronize_self_with_underlying_data_context()
+
+        self._variables = self._init_variables()
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/base_data_context.py:312]: Ah I see - this inherits `EphemeralDataContext._init_variables()`

File: great_expectations/data_context/data_context/base_data_context.py [hunk 5]
```diff
@@ -316,12 +319,12 @@ def __init__(
         # NOTE - 20210112 - Alex Sherstinsky - Validation Operators are planned to be deprecated.
         if (
             "validation_operators" in self.get_config().commented_map
-            and self.config.validation_operators
+            and self._project_config.validation_operators
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/base_data_context.py]: Here I think we still want to use the public property `self.config` rather than the private - or is there a reason I'm not thinking of?

File: great_expectations/data_context/data_context/base_data_context.py [hunk 6]
```diff
@@ -316,12 +319,12 @@ def __init__(
...
         ):
             for (
                 validation_operator_name,
                 validation_operator_config,
-            ) in self.config.validation_operators.items():
+            ) in self._project_config.validation_operators.items():
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/base_data_context.py]: Same comment (use self.config)

File: great_expectations/data_context/data_context/cloud_data_context.py [hunk 7]
```diff
@@ -98,7 +109,7 @@ def get_config_with_variables_substituted(
         be optional in GE Cloud mode).
         """
         if not config:
-            config = self.config
+            config = self._project_config
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/cloud_data_context.py]: Same question here, why not use the public self.config accessor?

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 2 above)
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py:254]: Once we move to Python 3.8 this might be a good candidate for a TypedDict
https://docs.python.org/3/library/typing.html#typing.TypedDict

Edit:
@cdkini pointed this out to me, no need for a TypedDict.
https://github.com/great-expectations/great_expectations/blob/feature/great-889/great-937/init-variables-in-datacontext/great_expectations/data_context/types/base.py

File: great_expectations/data_context/data_context/file_data_context.py [hunk 8]
```diff
@@ -98,7 +99,12 @@ def root_directory(self) -> Optional[str]:
         return self._context_root_directory
 
     def _init_variables(self) -> FileDataContextVariables:
-        raise NotImplementedError
+        variables: FileDataContextVariables = FileDataContextVariables(
+            config=self._project_config,
+            substitutions={},
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/file_data_context.py]: nit: For consistency since this is set to {} on init if not provided can you remove this line?

File: great_expectations/data_context/data_context_variables.py [hunk 9]
```diff
@@ -96,7 +97,7 @@ def _set(self, attr: DataContextVariableSchema, value: Any) -> None:
     def _get(self, attr: DataContextVariableSchema) -> Any:
         key: str = attr.value
         val: Any = self.config[key]
-        substituted_val: Any = substitute_config_variable(val, self.substitutions)
+        substituted_val: Any = substitute_all_config_variables(val, self.substitutions)
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context_variables.py:100]: Good catch

File: great_expectations/render/renderer/v3/suite_edit_notebook_renderer.py [hunk 10]
```diff
@@ -106,9 +105,7 @@ def __init__(
     def from_data_context(data_context: DataContext):
         suite_edit_notebook_config: Optional[NotebookConfig] = None
         if data_context.notebooks and data_context.notebooks.get("suite_edit"):
-            suite_edit_notebook_config = notebookConfigSchema.load(
-                data_context.notebooks.get("suite_edit")
-            )
+            suite_edit_notebook_config = data_context.notebooks.get("suite_edit")
```
Comment (anthonyburdi) [inline great_expectations/render/renderer/v3/suite_edit_notebook_renderer.py:108]: Why was this removed?

File: great_expectations/render/renderer/suite_edit_notebook_renderer.py [hunk 11]
```diff
@@ -90,9 +89,7 @@ def __init__(
     def from_data_context(data_context):
         suite_edit_notebook_config: Optional[NotebookConfig] = None
         if data_context.notebooks and data_context.notebooks.get("suite_edit"):
-            suite_edit_notebook_config = notebookConfigSchema.load(
-                data_context.notebooks.get("suite_edit")
-            )
+            suite_edit_notebook_config = data_context.notebooks.get("suite_edit")
```
Comment (anthonyburdi) [inline great_expectations/render/renderer/suite_edit_notebook_renderer.py:92]: Why was this removed?

**Review by anthonyburdi — CHANGES_REQUESTED** [review]
Some comments for discussion, will approve after resolving

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 3 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py:359]: Per our discussion, this isn't an issue but we may want to refactor to make this a bit cleaner. @Kilo59 had a good suggestion about context managers we may want to look into.

File: great_expectations/data_context/data_context/file_data_context.py (hunk 8 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/file_data_context.py]: Deleted!

**Review by Kilo59 — APPROVED** [review]

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py:254]: But a good suggestion elsewhere for sure 😄

File: great_expectations/render/renderer/v3/suite_edit_notebook_renderer.py (hunk 10 above)
Comment (cdkini) [reply great_expectations/render/renderer/v3/suite_edit_notebook_renderer.py:108]: Object is already in config form at this point due to changes made in variables

File: great_expectations/render/renderer/suite_edit_notebook_renderer.py (hunk 11 above)
Comment (cdkini) [reply great_expectations/render/renderer/suite_edit_notebook_renderer.py:92]: Object is already in config form at this point due to changes made in variables

File: great_expectations/data_context/data_context/cloud_data_context.py (hunk 7 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/cloud_data_context.py]: done!

File: great_expectations/data_context/data_context/base_data_context.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/base_data_context.py]: done!

File: great_expectations/data_context/data_context/base_data_context.py (hunk 5 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/base_data_context.py]: done!

**Review by Kilo59 — APPROVED** [review]

File: great_expectations/data_context/data_context/base_data_context.py [hunk 12]
```diff
@@ -384,7 +387,7 @@ def add_store(self, store_name: str, store_config: dict) -> Optional[Store]:
             store (Store)
         """
 
-        self.config["stores"][store_name] = store_config
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/base_data_context.py:387]: Oooh good catch. Can we still access via ["stores"]? I think so because DataContextConfig is a subclass of DictDot. If not this would be a breaking change.

**Review by anthonyburdi — APPROVED** [review]
LGTM! Please just double check config access via key still works e.g. `self.config["stores"]`

File: tests/conftest.py [hunk 13]
```diff
@@ -1507,10 +1507,12 @@ def site_builder_data_context_with_html_store_titanic_random(
     context.profile_datasource("random")
     context.profile_datasource(context.list_datasources()[0]["name"])
 
-    context._project_config.anonymous_usage_statistics = {
-        "enabled": True,
-        "data_context_id": "f43d4897-385f-4366-82b0-1a8eda2bf79c",
-    }
+    context.variables.config.anonymous_usage_statistics = (
```
Comment (donaldheppner) [inline tests/conftest.py]: why is this assigned through the config attribute and not `context.variables.anonymous_usage_statistics`?

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 13 above)
Comment (cdkini) [reply tests/conftest.py]: fixed!

**Review by donaldheppner — APPROVED** [review]

**Comment by Shinnnyshinshin** [comment]
since this PR still has some failing tests... along with #5472 I'm going to Disable auto-merge for now, so that we can get the other PRs in. Hope this is ok :)
